### ISSUE TITLE
Adaptive per-image pixel_budget: predict convergence from cheap first-pass stats

### ISSUE THREAD

**Issue by lilith** [imazen/zenanalyze#46]
## Goal

Today `DEFAULT_PIXEL_BUDGET` is one absolute cap (2 MP after #9 / PR #44). Every image pays the same upper bound regardless of whether the cheap budget already converged. Adaptive policy: do a first pass at default budget, look at one or two cheap-to-compute statistics, decide whether the picker can trust this image's features or whether a second pass at higher budget is worth it.

The shape:

```rust
let mut features = first_pass(slice, budget);
if needs_more_samples(&features) {
    extended_pass(slice, &mut features, budget * 2);  // re-run features that benefit
}
```

The second pass mutates the existing feature struct, costing extra compute only on the images that genuinely need it. The screen-content tail (the 6% pick disagreement at 2 MP we measured in the budget-research sweep) is the obvious target — those images should pay for the extra budget; photos shouldn't.

## Candidate convergence predictors (cheap, first-pass-computable)

| Signal | What it measures | Cheap? |
|---|---|---|
| `variance_spread` (`max_var / mean_var` over 8×8 blocks) | Heterogeneity of detail across the image | yes — already computed |
| `patch_fraction_fast` | Fraction of duplicate blocks (high → fewer unique samples needed) | yes |
| Bootstrap split-half delta | Compute feature on odd-stride pixels vs even-stride pixels; `‖odd − even‖` is a per-feature standard-error estimate | medium — ~5 % first-pass overhead, requires implementing the split |
| `is_grayscale` (binary, early-exit) | Exits on first non-gray pixel; very cheap on colored content | yes — task #33 / commit 8d72216 |
| `stride_step` at the picked budget | `pixel_count / budget` — if this is ~1, image is already exhaustive; if 16, sparse | yes — derive from inputs |
| `aq_map_p10` (flat-block count proxy) | Many flat blocks → those add no signal regardless of sample size | yes |

**`grayscale_score` is NOT a candidate.** Per the #11 perf audit, it's the heaviest unique-cost feature at 6533 µs / 4 MP. Use `is_grayscale` (the bool variant from #33) instead — same content distinction, ~1000× cheaper because it early-exits on the first non-gray pixel.

## Validation procedure

The existing 30-image, 5-budget sweep at `/mnt/v/output/zenpicker/budget-research-2026-04-30/picks_stability.tsv` already records per-image pick agreement at each of {0.5, 1, 2, 4, 16 MP}. Procedure:

1. **Bin images by `agreement(0.5 MP, 2 MP)`** — stable vs unstable.
2. **For each candidate predictor**, compute AUC for separating stable from unstable images using only the first-pass stats.
3. **Threshold tuning**: pick the predictor + threshold that maximizes precision at recall ≥ 0.85 — we want to catch most unstable images while not over-sampling stable ones.
4. **Composite predictor**: try `variance_spread > t1 OR patch_fraction_fast < t2` etc. — small logistic regression on the predictors.

Expected outcome: a single threshold or 2-feature gate that flags 80-90 % of unstable images (vs the 6 % baseline) and false-positives <20 % of the time on stable images. Net cost: extra pass on ~25 % of images instead of every image — wins on average.

## What's missing from the existing data (gating items)

- Only 30 images — small for AUC stability.
- Mostly photos — the unstable tail is dominated by 8 gb82-sc screen-content images. Need more screen-content + line-art + mixed-content for a robust predictor.
- Per-image **per-feature delta** between budgets isn't recorded, only pick agreement. Adding it to the sweep would let us train per-feature convergence predictors instead of one global gate.

## Acceptance criteria for shipping the adaptive policy

- AUC ≥ 0.80 for the predictor on the held-out part of the validation corpus
- Net runtime: extra-pass cost ≤ 30 % of always-on extended budget
- Pick agreement: improves vs flat 2 MP cap on screen content, no regression on photos

## Pointers

- `/mnt/v/output/zenpicker/budget_generalization_report_2026-04-30.md` — the empirical study behind PR #44
- `/mnt/v/output/zenpicker/budget-research-2026-04-30/picks_stability.tsv` — per-image data
- imazen/zenanalyze#9 / PR #44 — the absolute-cap bump that motivated this
- imazen/zenanalyze#11 — perf audit confirming `grayscale_score` is too expensive for the predictor list

## Quick first step

Run the AUC analysis against existing data first (30-image sweep). If any single predictor clears AUC ≥ 0.75, file a follow-up implementing the adaptive policy. If they all hover at 0.5, expand the corpus before investing further. ~10 minute analysis.

**Comment by lilith** [comment]
## AUC analysis on the 30-image, 5-budget sweep

Data: `/mnt/v/output/zenpicker/budget-research-2026-04-30/`. Label: `unstable = (pick@0.5MP != pick@2MP)` per `(image, target_zq)` row. n=150 rows from 30 images x 5 target_zq; **n_unstable=9** (base rate 6.0%).

**Caveat:** `feat_aq_map_p10` and `feat_is_grayscale` are not in this TSV (added later); used `feat_aq_map_mean`/`feat_aq_map_std` and `feat_grayscale_score` as substitutes.

### Concentration

All 9 unstable rows live on 6 screen-content images: `imac_dark` (3/5), `imac_g3` (2/5), `codec_wiki`, `gmessages`, `gui`, `imessage` (1/5 each). Zero unstable rows on photographic content. Pattern matches the gb82-sc hypothesis — the actual filenames just differ.

### Per-feature AUC (best direction per feature)

| predictor | AUC | thr@recall>=0.85 | precision | FPR |
|---|---|---|---|---|
| stride_step (high) | **0.898** | 6.32 | 0.23 | 0.19 |
| aq_map_std (high) | 0.868 | 1.36 | 0.15 | 0.36 |
| line_art_score (high) | 0.868 | 0.087 | 0.23 | 0.19 |
| aq_map_mean (low) | 0.861 | 1.29 | 0.23 | 0.22 |
| patch_fraction_fast (high) | 0.857 | 0.80 | 0.23 | 0.22 |
| uniformity (high) | 0.715 | 0.72 | 0.12 | 0.47 |
| grayscale_score (high) | 0.711 | 0.17 | 0.11 | 0.44 |
| variance_spread (low) | 0.679 | 1.55 | 0.09 | 0.55 |
| edge_density (high) | 0.463 | -- | -- | -- |

Note the direction flip on `patch_fraction_fast`: in this corpus **high** patch fraction predicts instability, not low. Screen-content has many flat-patch regions, and those are where stride sampling at 0.5 MP misses fine UI details.

### Best composite (in-sample logistic regression, balanced)

- 2-combo: `patch_fraction_fast` + `stride_step` -> AUC=**0.947**
- 3-combo: `variance_spread` + `patch_fraction_fast` + `stride_step` -> AUC=**0.951**

In-sample on n=150 with 9 positives — overfit risk is real, treat composite gains as suggestive only.

### Verdict

**Real signal, needs corpus expansion before shipping.** Single-feature AUC of 0.898 (stride_step) and 0.868 (aq_map_std, line_art_score) sit right on the 0.80-0.90 boundary, but with only 9 unstable cases the AUC CI is wide and threshold selection is unstable (recall=0.85 = catching 8 of 9). Predictor direction on `patch_fraction_fast` flipped vs the prior — wants confirmation on more images before being baked into a `const`.

Recommended next step: expand to ~150 images with at least 30 screen-content/UI/line-art samples, re-run, then ship the adaptive trigger using `stride_step >= ~6` OR `aq_map_std` high as the gate (cheap, no extra pass needed since stride_step is just `pixels/500_000`).

**Comment by lilith** [comment]
## Corpus expansion ready (gating prereq for re-running this analysis)

imazen/zenanalyze#47 is curated and ready: 150 images at `/mnt/v/output/zenpicker/budget-research-corpus-2026-04-30/` with a 30/60/30/30 (photo/screen/mixed/lineart) split. The gating block on "corpus too small for stable AUC" is now removed.

Next step (separate downstream task, **not this issue**): point `zenjpeg/examples/zq_pareto_calibrate.rs --corpus <path>` at the new 150-image corpus, run a 5-budget sweep (0.5/1/2/4/16 MP × picker), and re-run the AUC analysis with stable confidence intervals. Expected: ≥30 unstable cases (vs the 9 we had on the old 30-image set) which puts AUC point estimates within ±0.05 confidence.

Tracking issues for budget findings:
- imazen/zenanalyze#47 — corpus expansion (DONE — agent curated)
- imazen/zenanalyze#48 — patch_fraction_fast retrain trigger (filed)
- imazen/zenanalyze#49 — per-feature minimum sample-count floor (filed)

The picker bake stays calibrated for the OLD 0.5 MP budget until a re-extract+retrain at 2 MP runs after #47's data lands. Per the budget-generalization report's 94% pick agreement finding, the v2.2-clean ship candidate is fine to ship at the bumped budget — but the next bake (v2.3 or whatever lands after #48) should be calibrated coherent with PR #44's defaults.

**Comment by lilith** [comment]
Measured this end-to-end on 2026-08-30. Full write-up: [`benchmarks/budget_scaling_2026-08-30.md`](https://github.com/imazen/zenanalyze/blob/main/benchmarks/budget_scaling_2026-08-30.md). Short version: **the data does not support building this**, and it also fills in the three "gating items" above.

### First, a stale premise

This issue says `DEFAULT_PIXEL_BUDGET` "is one absolute cap (2 MP after #9 / PR #44)". It is **500 000** today (`src/feature.rs:2080`, `src/tier1.rs:212`) — 4× smaller than the number the plan above is reasoning about. Anything derived from the 2 MP figure needs re-deriving.

### The gating items, now filled

The sweep this issue wanted (per-image **per-feature** delta between budgets, on more than 30 mostly-photo images) exists now as `examples/budget_drift_grid.rs`: 117 features × 5 sizes (64…4096²) × 6 content classes × 8 crops, real corpus only, with `pixel_budget` and `hf_max_blocks` varied **independently**.

### Three results that bear on the proposal

**1. The budgets do not bind below 1024² at all.** Every feature is bit-identical across every budget arm at 64² and 256². An adaptive policy has nothing to do there; the whole question is a ≥ 1 MP question.

**2. Convergence is not monotone, which is the assumption `budget * 2` rests on.** Testing each feature along the ladder of the knob that actually drives it (default → 1M → 2M → 4M → 8M → full for `pixel_budget`; 1k → 2k → 4k → 16k → 64k → full for `hf_max_blocks`), at 4096²:

| | count |
|---|--:|
| never move — safe at any budget | 52 |
| move, converge on **all** cells | **4** |
| move, converge on **some** cells | 59 |
| move, converge on **no** cells | 2 |

Only 4 of the 65 moving features get monotonically closer to the fully-sampled value as the budget rises, and 3 of those 4 have p95 drift of 0.000 σ — they barely move at all. The non-converging group is dominated by the percentile and max statistics (`noise_floor_*_p*`, `aq_map_p*`, `quant_survival_*`), which is structural: a p75 of a sampled population is not a monotone function of sample count, and no predictor can make it one. So "second pass at 2× budget" does not reliably land closer to the truth for the features that actually move.

**3. The payoff is ~0.01 % bytes, and the cost is 4–24×.** Running the real consumers over both regimes (the shipped `zenpicker::default_route` through its actual entry point, plus mirrors of the zenjpeg picker and zenwebp classifier):

| consumer | 1024² | 2048² | 4096² |
|---|--:|--:|--:|
| zenjpeg picker | 9.46 % | 5.73 % | **0.00 %** |
| meta-picker | 3.39 % | 0.69 % | 3.91 % |
| zenwebp classifier | 0 % | 0 % | 0 % |

The 0 % at 4096² is **not** robustness — the zenjpeg picker emits *one* constant cell there across all 1 152 (image × target) combinations (15 distinct cells at 1024², 3 at 2048², 1 at 4096²). That is its own bug, filed separately; it also means no sampling policy can improve anything at 4K.

Where picks do flip, the bake's own predicted byte penalty for the default-budget pick (its outputs are log-bytes, so this is the picker's own objective) is median 0.26 % at 1024² and 1.11 % at 2048². Only 34 % of flips move `subsampling`, the knob with real byte consequences — 60 of 5 760 decisions ≥ 1024², so expected corpus-wide saving ≈ 0.012 % bytes.

Against a measured cost of 4.08× at 1024², 11.8× at 2048², 23.7× at 4096² (16.1 ms → 381.6 ms, which is **2.7× a fast JPEG encode of the same bytes on the same host**).

### Also: the knob in the title is the wrong one

`hf_max_blocks` drives more drift than `pixel_budget` does — 39 vs 26 moving features, and **every one of the 25 largest drifts at 4096² is tier-3 block-cap driven**. If adaptive sampling is ever revisited, it should start there, not at `pixel_budget`.

### Suggestion

Close as measured-not-worth-building, or re-scope to `hf_max_blocks` at ≥ 1 MP only. The 6 % screen-content disagreement that motivated this reproduces (screen classes do flip more), but it is worth ~1 % predicted bytes on ~1 % of decisions, and the adaptive machinery would sit on top of a non-monotone convergence curve.

Caveat on the regret figure: it is the model's own opinion, and it is a model trained at the default depth being asked to price full-depth inputs — the off-distribution substitution `Select::Features` exists to prevent. A ground-truth encode-and-score would break that circularity and was not run; the reasoning for stopping is that it would have to be off by two orders of magnitude to change the conclusion.
